[BLOG] tpchgen-rs: World’s fastest open source TPCH data generator, written in Rust (#67)
apache/datafusion-site: content/blog/2025-04-10-fastest-tpch-generator.md

---
layout: post
title: tpchgen-rs World’s fastest open source TPC-H data generator, written in Rust
date: 2025-04-10
author: Andrew Lamb, Achraf B, and Sean Smith
categories: [performance]
---

<!--
{% comment %}
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

<style>
/* Table borders */
table, th, td {
  border: 1px solid black;
  border-collapse: collapse;
}
th, td {
  padding: 3px;
}
</style>

3 members of the [Apache DataFusion] community used Rust and open source

> development to build [tpchgen-rs], a fully open TPC-H data generator over 2 0x

Reviewer: ```suggestion
development to build [tpchgen-rs], a fully open TPC-H data generator over 20x
```

Author: 🤦

analytical engines such as DataFusion.

<img src="/blog/images/fastest-tpch-generator/parquet-performance.png" alt="Time to create TPC-H parquet dataset for Scale Factor  1, 10, 100 and 1000" width="80%" class="img-responsive">

**Figure 1**: Time to create TPC-H dataset for Scale Factor (see below) 1, 10,
100 and 1000 as 8 individual SNAPPY compressed parquet files using a 22 core GCP
VM with 88GB of memory. For Scale Factor(SF) 100 `tpchgen` takes 1 minute and 14 seconds and
[DuckDB] takes 17 minutes and 48 seconds. For SF=1000, `tpchgen` takes 10
minutes and 26 and uses about 5 GB of RAM at peak, and we could not measure
DuckDB’s time as it [requires 647 GB of RAM], more than the 88 GB that was
available on our test machine. The testing methodology is in the
[documentation].

[DuckDB]: https://duckdb.org
[requires 647 GB of RAM]: https://duckdb.org/docs/stable/extensions/tpch.html#resource-usage-of-the-data-generator
[documentation]: https://github.com/clflushopt/tpchgen-rs/blob/main/benchmarks/BENCHMARKS.md

This blog explains what TPC-H is, how we ported the vintage C data generator to
Rust (yes, [RWIR]) and optimized its performance over the course of a few weeks

> of part-time work. We began this project so we can easily generate TPC-H data in

Reviewer 2: ```suggestion
of part-time work. We began this project so we could easily generate TPC-H data in
```

[Apache DataFusion] and [GlareDB].

[RWIR]: https://www.reddit.com/r/rust/comments/4ri2gn/riir_rewrite_it_in_rust/
[Apache DataFusion]: https://datafusion.apache.org/
[GlareDB]: https://glaredb.com/

# Try it for yourself

The tool is entirely open source under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0). Visit the [tpchgen-rs repository](https://github.com/clflushopt/tpchgen-rs) or try it for yourself by run the following commands after [installing Rust](https://www.rust-lang.org/tools/install):

```shell
$ cargo install tpchgen-cli

# create SF=1 in classic TBL format
$ tpchgen-cli -s 1 

# create SF=10 in Parquet
$ tpchgen-cli -s 10 --format=parquet
```

# What is TPC-H / dbgen?

The popular [TPC-H] benchmark (often referred to as TPCH) helps evaluate the

> performance of database systems on [OLAP] queries*, *the kind used to build BI

Reviewer 3: Not sure why the asterix's are here.

availability of free and open source database systems, it is common for users to
run and verify TPC-H performance themselves.

TPC-H simulates a business environment with eight tables: `REGION`, `NATION`,
`SUPPLIER`, `CUSTOMER`, `PART`, `PARTSUPP`, `ORDERS`, and `LINEITEM`. These
tables are linked by foreign keys in a normalized schema representing a supply
chain with parts, suppliers, customers and orders. The benchmark itself is 22
SQL queries containing joins, aggregations, and sorting operations.

The queries run against data created with <code>[dbgen]</code>, a program
written in a pre [C-99] dialect, which generates data in a format called *TBL*
(example in Figure 2). `dbgen` creates data for each of the 8 tables for a
certain *Scale Factor*, commonly abbreviated as SF. Example Scale Factors and
corresponding dataset sizes are shown in Table 1. There is no theoretical upper
bound on the Scale Factor.

[TPC-H]: https://www.tpc.org/tpch/
[OLAP]: https://en.wikipedia.org/wiki/Online_analytical_processing
[well known]: https://www.vldb.org/pvldb/vol9/p204-leis.pdf
[Transaction Processing Performance Council]: https://www.tpc.org/

> [dbgen]: https://github.com/electrum/tpch-dbgen)

Reviewer: ```suggestion
[dbgen]: https://github.com/electrum/tpch-dbgen
```

[Andrew and Sean started helping on March 8, 2025]: https://github.com/clflushopt/tpchgen-rs/commit/9bb386a4c55b8cf93ffac1b98f29b5da990ee79e
[released version 0.1]: https://crates.io/crates/tpchgen/0.1.0

## Optimizing Single Threaded Performance

Archaf [completed the end to end conformance tests], to ensure correctness, and
an initial [cli check in] on March 15, 2025.

[completed the end to end conformance tests]: https://github.com/clflushopt/tpchgen-rs/pull/16
[cli check in]: https://github.com/clflushopt/tpchgen-rs/pull/12

On a Macbook Pro M3 (Nov 2023), the initial performance numbers were actually
slower than the original Java implementation which was ported 😭. This wasn’t
surprising since the focus of the first version was to get a byte of byte
compatible port, and knew about the performance shortcomings and how to approach
them.

<table>
  <tr>
   <td><strong>Scale Factor</strong>
   </td>
   <td><strong>Time</strong>
   </td>
  </tr>
  <tr>
   <td>1
   </td>
   <td>0m10.307s
   </td>
  </tr>
  <tr>
   <td>10
   </td>
   <td>1m26.530s
   </td>
  </tr>
  <tr>
   <td>100
   </td>
   <td>14m56.986s
   </td>
  </tr>

> </table>

Reviewer 2: Just out of curiosity, why can't these be normal markdown tables?

Author: The practical reason is that this is how google docs --> markdown exported them

I actually tried to to make them normal markdown tables and the pelican renderer didn't seem to work

[several]: https://github.com/clflushopt/tpchgen-rs/issues/73 
[bugs]: https://github.com/clflushopt/tpchgen-rs/issues/65
[made]: https://github.com/clflushopt/tpchgen-rs/pull/71
[tpchgen-arrow]: https://crates.io/crates/tpchgen-arrow
[Apache Arrow]: https://arrow.apache.org/
[a small 300 line PR]: https://github.com/clflushopt/tpchgen-rs/pull/61
[a small 300 line PR]: https://github.com/clflushopt/tpchgen-rs/pull/61
[Rust Parquet writer]: https://crates.io/crates/parquet

<table>
  <tr>
   <td><strong>Scale Factor</strong>
   </td>
   <td><strong>Time to generate Parquet</strong>
   </td>
   <td><strong>Speed compared to tbl generation</strong>
   </td>
  </tr>
  <tr>
   <td>1
   </td>
   <td>0m1.649s
   </td>
   <td>0.8x
   </td>
  </tr>
  <tr>
   <td>10
   </td>
   <td>0m5.643s
   </td>
   <td>0.7x
   </td>
  </tr>
  <tr>
   <td>100
   </td>
   <td>0m45.243s
   </td>
   <td>0.7x
   </td>
  </tr>
  <tr>
   <td>100 (to /dev/null)
   </td>
   <td>0m45.153s
   </td>
   <td>0.5x
   </td>
  </tr>
</table>

**Table 5**: `tpchgen-cli` Parquet generation performance measured with  `time
target/release/tpchgen-cli -s $SCALE_FACTOR --format=parquet`

# Conclusion 👊🎤

> With a few days, some fellow database nerds, and the power of Rust we made
> something 10x better than currently exists. We hope it inspires more research

Reviewer 2: ```suggestion
In just a few days, with some fellow database nerds and the power of Rust, we built something 10x better than what currently exists. We hope it inspires more research
```

Flows a tad bit better as the start of a conclusion.

this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

<style>
/* Table borders */
table, th, td {
  border: 1px solid black;
  border-collapse: collapse;
}
th, td {
  padding: 3px;
}
</style>

3 members of the [Apache DataFusion] community used Rust and open source
development to build [tpchgen-rs], a fully open TPC-H data generator over 10x
faster than any other implementation  we know of.

It is now possible to create the TPC-H SF=100 dataset in 72.23 seconds (1.4 GB/s
😎) on a Macbook Air M3 with 16GB of memory, compared to the classic `dbgen`

> which takes 30 minutes<sup>1</sup> (0.05GB/sec). On the same machine, it takes less than

Reviewer: throughput 1.4GB/s vs 0.05GB/sec = 28x

Author: That is a good point -- the 10x is a conservative estimate.

It @clflushopt 's measurements (2m vs 44m is also 22x 🤔 )

The 10x is the smallest improvement on the gcp measurements (where the machine has more resources): https://docs.google.com/spreadsheets/d/14qTHR5zgqXq4BkhO1IUw2BPwBUIOqMXLZ2fUyOaPflI/edit?gid=0#gid=0

I'll update the text to say over 20x faster

> 2 minutes to create all 3.6 GB of SF=100 in [Apache Parquet] format.

Author: @clflushopt,

When reading this introduction I think it would also be nice to report the time taken to create Scale Factor 100 using DuckDB (my laptop has more memory and I think the numbers with more modest specs are more compelling)

 Is there any chance you can time how long it takes to run this script on your machine?

```sql
INSTALL tpch;
LOAD tpch;

.open test
.timer on
CALL dbgen(sf = 100);

-- Export each table to Parquet format
copy customer to 'customer.parquet' (FORMAT parquet);
copy lineitem to 'lineitem.parquet' (FORMAT parquet);
copy nation to 'supplier.parquet' (FORMAT parquet);
copy orders to 'supplier.parquet' (FORMAT parquet);
copy part to 'part.parquet' (FORMAT parquet);
copy partsupp to 'partsupp.parquet' (FORMAT parquet);
copy region to 'region.parquet' (FORMAT parquet);
copy supplier to 'supplier.parquet' (FORMAT parquet);
```

Reviewer 4: ```
~/G/P/tpchgen-rs (cl/feat/bump-crates-to-0.1.1)> duckdb
v1.2.0 5f5512b827
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D INSTALL tpch;
D LOAD tpch;
D
D .open test
D .timer on
D CALL dbgen(sf = 100);

-- Export each table to Parquet format
copy customer to 'customer.parquet' (FORMAT parquet);
copy lineitem to 'lineitem.parquet' (FORMAT parquet);
copy nation to 'supplier.parquet' (FORMAT parquet);
copy orders to 'supplier.parquet' (FORMAT parquet);
copy part to 'part.parquet' (FORMAT parquet);
copy partsupp to 'partsupp.parquet' (FORMAT parquet);
copy region to 'region.parquet' (FORMAT parquet);
copy supplier to 'supplier.parquet' (FORMAT parquet);

100%▕████████████████████████████
┌─────────┐
  │ Success │
  │ boolean │
 ├─────────┤
  │ 0 rows  │
└─────────┘
Run Time (s): real 687.659 user 1040.134996 sys 83.434645
D
D -- Export each table to Parquet format
D copy customer to 'customer.parquet' (FORMAT parquet);
100%▕███████████████████████████
Run Time (s): real 2.985 user 9.007536 sys 2.676405
D copy lineitem to 'lineitem.parquet' (FORMAT parquet);
100% ▕█████████████████████████████
Run Time (s): real 108.328 user 395.075322 sys 90.606267
D copy nation to 'supplier.parquet' (FORMAT parquet);
Run Time (s): real 0.010 user 0.002566 sys 0.001946
D copy orders to 'supplier.parquet' (FORMAT parquet);
100%▕███████████████████████████████
Run Time (s): real 19.461 user 78.310336 sys 17.984476
D copy part to 'part.parquet' (FORMAT parquet);
100%▕███████████████████████████████
Run Time (s): real 10.717 user 13.437501 sys 37.076887
D copy partsupp to 'partsupp.parquet' (FORMAT parquet);
100%▕████████████████████████████████
Run Time (s): real 11.286 user 31.645645 sys 14.751533
D copy region to 'region.parquet' (FORMAT parquet);
Run Time (s): real 0.003 user 0.000454 sys 0.000350
D copy supplier to 'supplier.parquet' (FORMAT parquet);
Run Time (s): real 0.500 user 0.613920 sys 0.130899
```

Reviewer 4: A less detailed more overview of the timing by running it as a script

```

jmp@comet ~/G/P/tpchgen-rs (main) [1]> time duckdb -init bench.sql -no-stdin
-- Loading resources from bench.sql
100% ▕████████████████████████████████████████████████████████████▏
┌─────────┐
│ Success │
│ boolean │
├─────────┤
│ 0 rows  │
└─────────┘
Run Time (s): real 717.838 user 1787.775069 sys 83.976228
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 5.471 user 11.592328 sys 12.049458
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 1015.519 user 456.350229 sys 104.793712
Run Time (s): real 0.010 user 0.002420 sys 0.002541
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 921.422 user 75.619220 sys 21.138167
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 2.647 user 12.597744 sys 1.699293
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 17.637 user 38.532114 sys 52.235873
Run Time (s): real 0.109 user 0.000459 sys 0.000680
Run Time (s): real 0.300 user 0.610369 sys 0.809258

________________________________________________________
Executed in   44.71 mins    fish           external
   usr time   39.72 mins    0.16 millis   39.72 mins
   sys time    4.63 mins    1.44 millis    4.63 mins
```

[RWIR]: https://www.reddit.com/r/rust/comments/4ri2gn/riir_rewrite_it_in_rust/
[Apache DataFusion]: https://datafusion.apache.org/
[GlareDB]: https://glaredb.com/

# Try it for yourself

The tool is entirely open source under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0). Visit the [tpchgen-rs repository](https://github.com/clflushopt/tpchgen-rs) or try it for yourself by run the following commands after [installing Rust](https://www.rust-lang.org/tools/install):

```shell
$ cargo install tpchgen-cli

# create SF=1 in classic TBL format
$ tpchgen-cli -s 1 

# create SF=10 in Parquet
$ tpchgen-cli -s 10 --format=parquet
```

# What is TPC-H / dbgen?

The popular [TPC-H] benchmark (often referred to as TPCH) helps evaluate the
performance of database systems on [OLAP] queries*, *the kind used to build BI
dashboards.

TPC-H has become a de facto standard for analytic systems. While there are [well
known] limitations as the data and queries do not well represent many real world
use cases, the majority of analytic database papers and industrial systems still
use TPC-H query performance benchmarks as a baseline. You will inevitably find

> multiple results for  “`TPCH Performance &lt;your favorite database>`” in any

Reviewer: ```suggestion
multiple results for  “`TPCH Performance <your favorite database>`” in any
```

3. It outputs TBL format, which typically requires loading into your database (for example, [here is how to do so] in Apache DataFusion) prior to query.
4. The implementation makes substantial assumptions about the operating environment, making it difficult to extend or embed into other systems.<sup>2</sup>

[electrum/tpch-dbgen]: https://github.com/electrum/tpch-dbgen
[here is how to do so]: https://github.com/apache/datafusion/blob/507f6b6773deac69dd9d90dbe60831f5ea5abed1/datafusion/sqllogictest/test_files/tpch/create_tables.slt.part#L24-L124
 

<img src="/blog/images/fastest-tpch-generator/tbl-performance.png" alt="Time to generate TPC-H data in TBL format" width="80%" class="img-responsive">

**Figure 3**: Time to generate TPC-H data in TBL format. The default `tpchgen` is
shown in blue. `tpchgen` restricted to a single core is shown in red. Unmodified
`dbgen` is shown in green and `dbgen` modified to use `-O3` optimization level
is shown in yellow.

`dbgen` is so inconvenient and takes so long that vendors often provide
preloaded TPC-H data, for example [Snowflake Sample Data], [DataBricks Sample
datasets] and [DuckDB Pre-Generated Data Sets].

> [Snowflake Sample  Data]: https://docs.snowflake.com/en/user-guide/sample-data-tpch

Reviewer: hyperlink not generated 

![Screenshot 2025-04-05 at 11 12 38 AM](https://github.com/user-attachments/assets/004cd640-9c44-4682-83cb-af1296d0065e)

```suggestion
[Snowflake Sample Data]: https://docs.snowflake.com/en/user-guide/sample-data-tpch
```

> [DataBricks Sample datasets]: https://docs.databricks.com/aws/en/discover/databricks-datasets

Reviewer: nit:
`DataBricks` -> `Databricks`

if changed, also change the reference above for hyperlink to work

[DuckDB Pre-Generated Data Sets]: https://duckdb.org/docs/stable/extensions/tpch.html#pre-generated-data-sets

In addition to pre-generated datasets, DuckDB also provides a [TPCH extension] 
for generating TPC-H datasets within DuckDB. This is so much easier to use than
the current alternatives that it leads many researchers and other thought
leaders to use DuckDB to evaluate new ideas. For example, [Wan Shen
Lim] explicitly [mentioned the ease of creating the TPC-H dataset] as one reason
the first student project of [CMU-799 Spring 2025] used DuckDB.

> [TPC-H extension]: https://duckdb.org/docs/stable/extensions/tpch.html

Reviewer: looking at the rest of the blog, maybe `TPC-H` is right. so we should change the sentence above

   </td>
  </tr>
</table>

**Table 3**: Single threaded `tpchgen-cli` performance, measured with `time target/release/tpchgen-cli -s $SCALE_FACTOR --num-threads=1`

## Multi-threading

Then we applied [Rust’s fearless concurrency] – with a single, [small PR] (272
net new lines) we updated the same memory safe code to run with multiple threads
and consume bounded memory using [tokio for the thread scheduler]<sup>4</sup>.

[Rust’s fearless concurrency]: https://doc.rust-lang.org/book/ch16-00-concurrency.html 
[small PR]: https://github.com/clflushopt/tpchgen-rs/commit/ab720a70cdc80a711f4a3dda6bac05445106f499
[tokio for the thread scheduler]: https://thenewstack.io/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/

As shown in Table 4, with this change, tpchgen-cli generates the full SF=100
dataset in 32 seconds (which is 3.3 GB/sec 🤯). Further investigation reveals
that at SF=100 our generator is actually IO bound (which is not the case for
`dbgen` or `duckdb`) – it creates data **faster than can be written to an SSD**.
When writing to `/dev/null` tpchgen  generates the entire dataset in 25 seconds
(4 GB/s).

<table>
  <tr>
   <td><strong>Scale Factor</strong>
   </td>

>    <td>Time

Reviewer: nit: same style as other columns
```suggestion
   <td><strong>Time</strong>
```

needed for the TPC-H benchmarks (e.g. floating point vs Decimal).

What would be far more useful is a typed, efficient columnar format such as
Apache Parquet which is supported by all modern query engines. So we [made] a
[tpchgen-arrow] crate to create [Apache Arrow] arrays directly and then [a small
300 line PR] to feed those arrays to the [Rust Parquet writer], again using
tokio for parallelized but memory bound work.

This approach was simple, fast and scalable, as shown in Table 5. Even though
creating Parquet files is significantly more computationally expensive than TBL
or CSV, tpchgen-cli creates the full SF=100 parquet format dataset in less than
45 seconds.

[added support for CSV]: https://github.com/clflushopt/tpchgen-rs/pull/54
[@niebayes]: https://github.com/niebayes
[fixing]: https://github.com/clflushopt/tpchgen-rs/pull/66 
[several]: https://github.com/clflushopt/tpchgen-rs/issues/73 
[bugs]: https://github.com/clflushopt/tpchgen-rs/issues/65
[made]: https://github.com/clflushopt/tpchgen-rs/pull/71
[tpchgen-arrow]: https://crates.io/crates/tpchgen-arrow
[Apache Arrow]: https://arrow.apache.org/
[a small 300 line PR]: https://github.com/clflushopt/tpchgen-rs/pull/61

> [Apache Arrow](https://arrow.apache.org/

Reviewer: redundant
![Screenshot 2025-04-05 at 11 19 13 AM](https://github.com/user-attachments/assets/ffde28bb-117e-4af8-8be1-1ce830d4c54d)

```suggestion
```

[a small 300 line PR]: https://github.com/clflushopt/tpchgen-rs/pull/61
[Rust Parquet writer]: https://crates.io/crates/parquet

<table>
  <tr>

>    <td>Scale Factor
>    </td>
>    <td>Time to generate Parquet
>    </td>
>    <td>Speed compared to tbl generation

Reviewer: nit: same style as other tables
```suggestion
   <td><strong>Scale Factor</strong>
   </td>
   <td><strong>Time to generate Parquet</strong>
   </td>
   <td><strong>Speed compared to tbl generation</strong>
```

   </td>
  </tr>

>   <tr>
>    <td>1
>    </td>
>    <td>0m1.649s
>    </td>
>    <td>0.8
>    </td>
>   </tr>
>   <tr>
>    <td>10
>    </td>
>    <td>0m5.643s
>    </td>
>    <td>0.7
>    </td>
>   </tr>
>   <tr>
>    <td>100
>    </td>
>    <td>0m45.243s
>    </td>
>    <td>0.7
>    </td>
>   </tr>
>   <tr>
>    <td>100 (to /dev/null)
>    </td>
>    <td>0m45.153s
>    </td>
>    <td>0.5
>    </td>
>   </tr>

Reviewer: nit:
```suggestion
  <tr>
   <td>1
   </td>
   <td>0m1.649s
   </td>
   <td>0.8x
   </td>
  </tr>
  <tr>
   <td>10
   </td>
   <td>0m5.643s
   </td>
   <td>0.7x
   </td>
  </tr>
  <tr>
   <td>100
   </td>
   <td>0m45.243s
   </td>
   <td>0.7x
   </td>
  </tr>
  <tr>
   <td>100 (to /dev/null)
   </td>
   <td>0m45.153s
   </td>
   <td>0.5x
   </td>
  </tr>
```

---
layout: post

> title: `tpchgen-rs` World’s fastest open source TPCH data generator, written in Rust

Author: fixed

date: 2025-04-10
author: Andrew Lamb, Achraf B, and Sean Smith
categories: [performance]
---

<!--
{% comment %}
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

<style>
/* Table borders */
table, th, td {
  border: 1px solid black;
  border-collapse: collapse;
}
th, td {
  padding: 3px;
}
</style>

3 members of the [Apache DataFusion] community used Rust and open source

> development to build [tpchgen-rs], a fully open TPCH data generator over 10x

Reviewer 5: According to the specification, it is `TPC-H` rather than `TPCH`.

Author: Thank you -- updated in 4eec529

faster than any other implementation  we know of.

It is now possible to create the TPCH SF=100 dataset in 72.23 seconds (1.4 GB/s
😎) on a Macbook Air M3 with 16GB of memory, compared to the classic `dbgen`
which takes 30 minutes[^1] (0.05GB/sec). On the same machine, it takes less than
2 minutes to create all 3.6 GB of SF=100 in [Apache Parquet] format.

> Finally, it is convenient and efficient to run TPCH queries locally when testing

Author: Updated in 12df7fa

   </td>
  </tr>
  <tr>
   <td>0.1
   </td>
   <td>103 Mb
   </td>
   <td>31 Mb
   </td>
  </tr>
  <tr>
   <td>1
   </td>
   <td>1 Gb
   </td>
   <td>340 Mb
   </td>
  </tr>
  <tr>
   <td>10
   </td>
   <td>10 Gb
   </td>
   <td>3.6 Gb
   </td>
  </tr>
  <tr>
   <td>100
   </td>
   <td>107 Gb
   </td>
   <td>38 Gb
   </td>
  </tr>
  <tr>
   <td>1000
   </td>
   <td>1089 Gb
   </td>
   <td>379 Gb
   </td>
  </tr>
</table>

**Table 1**: TPCH data set sizes at different scale factors for both TBL and [Apache Parquet].

[Apache Parquet]: https://parquet.apache.org/

# Why do we need a new TPCH Data generator?

Despite the known limitations of the TPCH benchmark, it is so well known that it
is used frequently in database performance analysis. To run TPCH, you must first
load the data, using `dbgen`, which is not ideal for several reasons:

1. You must find and compile a copy of the 15+ year old C program (for example [electrum/tpch-dbgen])
2. `dbgen` requires substantial time (Figure 3) and is not able to use more than one core.
3. It outputs TBL format, which typically requires loading into your database (for example, [here is how to do so] in Apache DataFusion) prior to query.

> 4. The implementation makes substantial assumptions about the operating environment, making it difficult to extend or embed into other systems.[^2]

Author: I reworked the footnotes in 4dc8aa0

into analytical systems using the TPCH dataset and that people build awesome
things with it. For example, Sean has already added [on-demand generation of
tables to GlareDB]. Please consider joining us and helping out at
[https://github.com/clflushopt/tpchgen-rs].

We met while working together on Apache DataFusion in various capacities. If you
are looking for a community of like minded people hacking on databases, we
welcome you to [come join us]. We are in the process of integrating this into
DataFusion (see [apache/datafusion#14608]) if you are interested in helping 🎣

[on-demand generation of tables to GlareDB]: https://github.com/GlareDB/glaredb/pull/3549
[https://github.com/clflushopt/tpchgen-rs]: https://github.com/clflushopt/tpchgen-rs
[come join us]: https://datafusion.apache.org/contributor-guide/communication.html
[apache/datafusion#14608]: https://github.com/apache/datafusion/issues/14608

# About the Authors:

- [Andrew Lamb] ([@alamb]) is a Staff Engineer at [InfluxData] and a PMC member of [Apache DataFusion] and [Apache Arrow].
- Achraf B ([@clflushopt]) is a Software Engineer at [Optable] where he works on data infrastructure.

> - [Sean Smith] ([@scsmithr]) is the founder of  focused on building a fast analytics database.

Reviewer 6: ```suggestion
- [Sean Smith] ([@scsmithr]) is the founder of [GlareDB] focused on building a fast analytics database.
```

Reviewer 6: Correct me if I'm wrong, based on the LinkedIn link :-)

Author: Yes for sure -- great catch

> [tpchgen-rs]: https://github.com/alamb/tpchgen-rs

Reviewer 6: Should we use this link: https://github.com/clflushopt/tpchgen-rs?

Author: Yes, also a great catch

```suggestion
[tpchgen-rs]: https://github.com/clflushopt/tpchgen-rs
```
